### PR TITLE
Move all waiting list related logic from staking.go

### PR DESCRIPTION
**Pull request by mariusmihaic**

### CODE REVIEW

File: epochStart/metachain/systemSCs.go [hunk 1]
```diff
@@ -242,7 +242,7 @@ func (s *systemSCProcessor) ProcessSystemSmartContract(
 		}
 	}
 
-	if s.flagCorrectLastUnjailedEnabled.IsSet() {
```
[the remark this answers is not in the document]
Comment (mariusmihaic) [reply epochStart/metachain/systemSCs.go:245]: Good point, removed it

File: epochStart/metachain/systemSCs.go [hunk 2]
```diff
@@ -256,14 +256,14 @@ func (s *systemSCProcessor) ProcessSystemSmartContract(
...
-		err := s.cleanAdditionalQueue() // TODO: Deactivate this?
+	if s.flagCorrectNumNodesToStake.IsSet() && !s.flagStakingV4Enabled.IsSet() {
+		err := s.cleanAdditionalQueue()
 		if err != nil {
 			return err
 		}
 	}
 
-	if s.flagSwitchJailedWaiting.IsSet() {
+	if s.flagSwitchJailedWaiting.IsSet() && !s.flagStakingV4Enabled.IsSet() {
```
[the remark this answers is not in the document]
Comment (mariusmihaic) [reply epochStart/metachain/systemSCs.go:266]: Right, good ideea, changed in `EpochConfirmed` using `stakingV4**Init**EnableEpoch` as following:
```go
s.flagSwitchJailedWaiting.SetValue(epoch >= s.switchEnableEpoch && epoch < s.stakingV4InitEnableEpoch)
```

File: vm/systemSmartContracts/staking.go [hunk 3]
```diff
@@ -1298,6 +1317,13 @@ func (s *stakingSC) getWaitingListIndex(args *vmcommon.ContractCallInput) vmcomm
 		return vmcommon.UserError
 	}
 
+	if s.flagStakingV4.IsSet() {
```
[the remark this answers is not in the document]
Comment (mariusmihaic) [reply vm/systemSmartContracts/staking.go:1320]: Right, moved it

File: vm/systemSmartContracts/staking.go [hunk 4]
```diff
@@ -588,11 +598,16 @@ func (s *stakingSC) unStakeAtEndOfEpoch(args *vmcommon.ContractCallInput) vmcomm
...
 	}
-	if registrationData.Waiting {
-		err = s.removeFromWaitingList(args.Arguments[0])
-		if err != nil {
-			s.eei.AddReturnMessage(err.Error())
-			return vmcommon.UserError
+
+	// This is an extra check. We should not save any registrationData
+	// with Waiting = true when staking v4 is enabled
+	if !s.flagStakingV4.IsSet() {
```
[the remark this answers is not in the document]
Comment (mariusmihaic) [reply vm/systemSmartContracts/staking.go:604]: Right, thought I'd add it as an extra safety check, but could be removed

File: vm/systemSmartContracts/staking.go [hunk 5]
```diff
@@ -1147,6 +1162,10 @@ func createWaitingListKey(blsKey []byte) []byte {
 }
 
 func (s *stakingSC) switchJailedWithWaiting(args *vmcommon.ContractCallInput) vmcommon.ReturnCode {
+	if s.flagStakingV4.IsSet() {
```
[the remark this answers is not in the document]
Comment (mariusmihaic) [reply vm/systemSmartContracts/staking.go:1165]: Added a check there;

Above that code, there are also these lines:

```go 
	if !registrationData.Staked {
		registrationData.Waiting = false
		err = s.removeFromWaitingList(args.Arguments[0])
		if err != nil {
			s.eei.AddReturnMessage(err.Error())
			return vmcommon.UserError
		}
		err = s.saveStakingData(args.Arguments[0], registrationData)
		if err != nil {
			s.eei.AddReturnMessage("cannot save staking data: error " + err.Error())
			return vmcommon.UserError
		}

		return vmcommon.Ok
	}
```

Should we also add a check for these?

File: vm/systemSmartContracts/staking.go [hunk 6]
```diff
@@ -1364,6 +1390,13 @@ func (s *stakingSC) getWaitingListSize(args *vmcommon.ContractCallInput) vmcommo
 		return vmcommon.OutOfGas
 	}
 
+	if s.flagStakingV4.IsSet() {
```
[the remark this answers is not in the document]
Comment (mariusmihaic) [reply vm/systemSmartContracts/staking.go:1393]: Moved it at the top of the file; will also create an extra file

File: vm/systemSmartContracts/staking.go [hunk 7]
```diff
@@ -1581,14 +1614,19 @@ func (s *stakingSC) getTotalNumberOfRegisteredNodes(args *vmcommon.ContractCallI
 		return vmcommon.UserError
 	}
 
-	stakeConfig := s.getConfig()
-	waitingListHead, err := s.getWaitingListHead()
-	if err != nil {
-		s.eei.AddReturnMessage(err.Error())
-		return vmcommon.UserError
+	waitingListLength := int64(0)
+	if !s.flagStakingV4.IsSet() {
```
[the remark this answers is not in the document]
Comment (mariusmihaic) [reply vm/systemSmartContracts/staking.go:1618]: Right, I was thinking the same. This flag could hide potential bugs

File: vm/systemSmartContracts/staking.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (sasurobert) [reply vm/systemSmartContracts/staking.go]: can stay.

[the remark this answers is not in the document]
Comment (sasurobert) [reply vm/systemSmartContracts/staking.go]: add error everywhere were we do no intend the code to be reached.

**Pull request closed without merging**